File: algoexpert/easy/javascript/Node_Depth.js
// Iterative approach
// O(n) time | O(1) space
const Node_Depth = root => {
	let sum = 0;
	let stack = [{"node": root, "depth": 0}];
	while (node.length > 0) {
		let nodeInfo = stack.pop();
		node = nodeInfo.node;
		depth = nodeInfo.depth;
		if (node === null)
			continue;
		sum += depth;
		stack.push({ "node": node.left, "depth": depth + 1 });
		stack.push({ "node": node.right, "depth": depth + 1 });
	}
	return sum;
}

// recursive approach
// O(n) time | O(h) space
const depth_node = (root, depth = 0) => {
	// handle the base case of recursion
	if (root === null) return 0;
	return depth + depth_node(root.left, depth + 1) + depth_node(root.right, depth + 1);
}

File: algoexpert/easy/python/Node_Depth.py
# Iterative approach
# O(n) time | O(1) space

def node_depth(root):
	sum = 0
	stack = [{"node": root, "depth": 0}]
	while len(stack) > 0:
		nodeInfo = stack.pop()
		node, depth = nodeInfo["node"], nodeInfo["depth"]
		if node is None:
			continue
		sum += depth
		stack.append([{"node": root.left, "depth": depth + 1}])
		stack.append([{"node": root.right, "depth": depth + 1}])
	return sum

# Recursive approach
# O(n) time | O(h) space

def depth_node(root, depth=0):
	if root is None:
		return 0
	return depth + depth_node(root.left, depth + 1) + depth_node(root.right, depth + 1)